### PR TITLE
Update group.partitions_for_topic to fetch topic metadata if it does …

### PR DESCRIPTION
**Pull request by Baisang**
…not have it

As mentioned in https://github.com/dpkp/kafka-python/issues/1774, this attempts to simulate the Java client, which does the following

```
public java.util.List<PartitionInfo> partitionsFor(java.lang.String topic)
Get metadata about the partitions for a given topic. This method will issue a remote call to the server if it does not already have any metadata about the given topic.
```
https://kafka.apache.org/10/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html#partitionsFor

---
This change is [Reviewable](https://reviewable.io/reviews/dpkp/kafka-python/1781)

### CODE REVIEW

File: kafka/consumer/group.py [hunk 1]
```diff
@@ -567,18 +564,34 @@ def topics(self):
...
         Arguments:
             topic (str): Topic to check.
 
         Returns:
             set: Partition ids
         """
-        return self._client.cluster.partitions_for_topic(topic)
+        cluster = self._client.cluster
+        partitions = cluster.partitions_for_topic(topic)
+        if not partitions:
```
Comment (jeffwidman) [inline kafka/consumer/group.py]: The return value of `partitions` is `None`, so let's explicitly check for that: `if partitions is None`

Additionally, I noticed that a non-existent topic will also return `None`, so this will issue a remote call if the topic simply does not exist... I can't think of any perf problems this will cause and it's  already mentioned in my suggested docstring so I think it's fine.

```
In [1]: from kafka import KafkaConsumer

In [2]: kc = KafkaConsumer('test_topic', group_id='testings', bootstrap_servers=['172.28.0.3'])

In [3]: cluster = kc._client.cluster

In [4]: cluster.partitions_for_topic('test_poll_group_by_topic_true')

In [5]: cluster.partitions_for_topic('non_existent_topic')

In [6]: kc.topics()
Out[6]: {u'test_poll_group_by_topic_true'}

In [7]: cluster.partitions_for_topic('non_existent_topic')

In [8]: cluster.partitions_for_topic('test_poll_group_by_topic_true')
Out[8]: {0}
```

File: kafka/consumer/group.py [hunk 2]
```diff
@@ -567,18 +564,34 @@ def topics(self):
         future = cluster.request_update()
         self._client.poll(future=future)
         cluster.need_all_topic_metadata = stash
+
+    def topics(self):
+        """Get all topics the user is authorized to view.
```
Comment (jeffwidman) [inline kafka/consumer/group.py]: Can you add:
`This will always issue a remote call to the cluster to fetch the latest information`.

File: kafka/consumer/group.py [hunk 3]
```diff
@@ -567,18 +564,34 @@ def topics(self):
...
+            set: topics
+        """
+        cluster = self._client.cluster
+        self._fetch_all_topic_metadata()
         return cluster.topics()
 
     def partitions_for_topic(self, topic):
         """Get metadata about the partitions for a given topic.
+        This method will issue a remote call to the server if it does not already
+        have any metadata about the given topic.
```
Comment (jeffwidman) [inline kafka/consumer/group.py]: Let's make it clear in the docstring that there will be times when this method could return values that do not match the state of the cluster due to an outdated cache entry... 

Also, while I have never played with Kafka ACL's, I'm guessing since `_fetch_all_topic_metadata()` only returns info about topics the user is authorized to view that trying to fetch info about a topic that exists but the user can't view will also trigger a remote call / returning `None`? Normally I'd expect a call about a specific topic to return some kind of error from the cluster about the user not authorized to view but here the cluster never sees a topic-scoped call so that's probably not the case.

maybe reword to:

`This method first checks the local metadata cache for information about the topic. If the topic is not found (either because the topic does not exist, the user is not authorized to view the topic, or the metadata cache is not populated), then it will issue a metadata update call to the cluster.`

File: kafka/consumer/group.py [hunk 4]
```diff
@@ -552,11 +552,8 @@ def committed(self, partition):
                 committed = None
         return committed
 
-    def topics(self):
-        """Get all topics the user is authorized to view.
-
-        Returns:
-            set: topics
+    def _fetch_all_topic_metadata(self):
```
Comment (jeffwidman) [inline kafka/consumer/group.py]: Is the Java client structured this way with this private method?

It's a nice clean way to do it this way, but just curious how they handle it...

File: kafka/consumer/group.py [hunk 5]
```diff
@@ -552,11 +552,8 @@ def committed(self, partition):
...
+        """A blocking call that fetches topic metadata for all topics in the cluster.
```
Comment (jeffwidman) [inline kafka/consumer/group.py]: I'm guessing the statement `...that the user is authorized to view.` also applies here?

File: kafka/consumer/group.py [hunk 6]
```diff
@@ -567,18 +564,34 @@ def topics(self):
...
         self._client.poll(future=future)
         cluster.need_all_topic_metadata = stash
+
+    def topics(self):
+        """Get all topics the user is authorized to view.
+
+        Returns:
+            set: topics
+        """
+        cluster = self._client.cluster
```
Comment (jeffwidman) [inline kafka/consumer/group.py]: Why is this needed?

Why not just change it to `return self._client.cluster.topics()`?

File: kafka/consumer/group.py (hunk 3 above)
Comment (Baisang) [reply kafka/consumer/group.py]: I also have not played with Kafka's ACL's, but that would be my assumption as well.

File: kafka/consumer/group.py (hunk 4 above)
Comment (Baisang) [reply kafka/consumer/group.py]: Yep! I think that's what I based it on https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L1892

**Comment by Baisang** [comment]
@jeffwidman Thanks for the feedback -- I addressed your comments, hopefully travis is kind to me.

**Comment by jeffwidman** [comment]
🎆 

Thanks again for this!
